doc/adr/0006-terraform-code-structure.md

```diff
@@ -0,0 +1,129 @@
+
+    An example may help clear this up.
+
+    Most of the DSO space is made up of a cluster of WebLogic servers speaking
+    to a number of Oracle Databases, with some Engineering tools around the
+    edges in an azure subscription.
+
+    It might make sense to have a single functional module to setup and
+    configure some common resources. For instance a base module may deploy a
+    resource group and within that configure a vnet, an application subnet, a
+    data subnet and an engineering tools subnet. Other potential resources could
+    be network security groups for each of the subnets.
+
+    With appropriate inputs to configure it, this functional module could then
+    be used by all of the application stacks DSO are responsible for.
+
+    All functional modules would also be required to expose relevant resources
+    via outputs. This would allow other modules to attach network rules to the
+    NSGs in the example above. This allows a relevant seperation of concerns.
+
+    A functional module may make use of serveral other terraform modules where
+    that makes sense.
+
+  - Rely more on modules inputs rather than remote state data.
+
+    Using Terraform v0.12 we can pass whole resources as variables so this
+    becomes much less tedious than in previous versions.
+
+    Currently existing terragrunt implementations do not utilise inputs to pass
+    data between modules. Instead as each module is invoked seperately, due to
+    the use of terragrunt, there are a number of state files used, and existing
+    modules make use of these. However this means each module that utilises the
+    remote state of another module tightly binds the two modules togeather.
+    As if the remote state changes, e.g. names of resources, then all modules
+    using that remote state may break with no warning.
+
+    If instead inputs are used, a module can require that there be subnet and
+    NSG resources passed in as inputs, so it can deploy servers and network
+    security rules. The module does not care how those resources are created,
+    and it becomes the calling modules responsibility to ensure that the correct
+    inputs are provided.
+
+## Consequences
+
+As a result of these changes our terraform code should become more composable
+and clearer. Functional modules will be reused and should be written in a fairly
+generic manner, while root modules will be written to use those functional
+modules to deploy a specific peice of infrastructure. For instance we may have a
+few functional modules, `azure_base_networking`, `azure_weblogic_oracle_db_app`,
+`azure_engineering_tools` modules, which with different inputs could be used by
+the following root modules, `azure_nomis`, `azure_mis`, `azure_oasys`.
+
+## Credit
+
+Thanks to @daibach and the LAA team at MoJ. This first ADR and the repo structure was copied from https://github.com/ministryofjustice/laa-hosting-architectural-decisions.
```

Comment:
These credit blocks can be removed from both ADRs. That was just in the first ADR because it was lifted pretty much verbatim from that repo.